### PR TITLE
Feature flow interface

### PR DESCRIPTION
**Pull request by lukanus**

### CODE REVIEW

**Review by christophermanning — APPROVED** [review]

File: client/client.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+		}
+
+		numberStr := strings.Replace(bH.Number, "0x", "", -1)
+		n := new(big.Int)
+		n.SetString(numberStr, 16)
+		hi.Height = n.Uint64()
+	}
+
+	params, err := sess.Send(ctx, []structs.Req{
+		{Method: MethodGetBlockHash, Params: []interface{}{hi.Height}},     // hash
```
Comment (deelawn) [inline client/client.go:71]: I'm not too familiar with Polkadot. Is it ever possible that we could reindex from height 0 or 1? If so this should be checked here before adding parent and grandparent.

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+		{Method: MethodGetStorage, Params: []interface{}{PolkadotTypeTimeNow, blockHash}},
+		{Method: MethodGetStorage, Params: []interface{}{PolkadotTypeSystemEvents, blockHash}},
+		{Method: MethodGetStorage, Params: []interface{}{PolkadotTypeNextFeeMultiplier, parentBlockHash}},
+		{Method: MethodGetStorage, Params: []interface{}{PolkadotTypeCurrentEra, parentBlockHash}},
+		{Method: "state_getRuntimeVersion", Params: []interface{}{grandParentBlockHash}},
+	}
+
+	resp, err := sess.Send(ctx, reqs)
+	if err != nil {
+		return ddr, block, prm, err
```
Comment (deelawn) [inline client/client.go:193]: If the return variables are named you can make it easier on yourself and simply do a `return`

File: client/client.go [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+	}
+
+	return ddr, block, prm, err
+}
+
+func respToVar(resp structs.Resp) ([]byte, error) {
+	if len(resp.Result) == 0 {
+		return []byte(""), nil
+	}
+	s := string(resp.Result)
```
Comment (deelawn) [inline client/client.go:253]: Is the purpose of this to make sure that the original slice is preserved and does not get changed?

**Review by deelawn — APPROVED** [review]
Looks good as it is.  I added a few comments, questions, and suggestions.

File: api/api.go [hunk 4]
```diff
@@ -102,73 +113,66 @@ func (conn *Conn) getStatus() ConnStatus {
 }
 
 // Send is there just because of mock, it doesn't make much sense otherwise
```
[the remark this answers is not in the document]
Comment (lukanus) [reply api/api.go:115]: Ups sorry :)

File: api/api.go [hunk 5]
```diff
@@ -102,73 +113,66 @@ func (conn *Conn) getStatus() ConnStatus {
...
 	var id uint64
-	for {
-		select {
-		case <-tckr.C:
-			status := conn.getStatus()
-			if status != StateOnline {
-				conn.l.Warn("[API] Connection is NOT Online")
-				continue
-			}
+	for _ = range tckr.C {
```
[the remark this answers is not in the document]
Comment (lukanus) [reply api/api.go:126]: :+1

File: api/api.go [hunk 6]
```diff
@@ -102,73 +113,66 @@ func (conn *Conn) getStatus() ConnStatus {
...
+		}
 
-	conn.Requests <- JsonRPCSend{
-		RespCH:         ch,
-		JsonRPCRequest: JsonRPCRequest{ID: id, Method: method, Params: params},
 	}
-	return nil
 }
 
 func (conn *Conn) recv(ctx context.Context, c *websocket.Conn, done chan struct{}, resps *LockedResponseMap) {
```
[the remark this answers is not in the document]
Comment (lukanus) [reply api/api.go:178]: Well, it is :) 
It's closing this channel after it's done, and close would trigger other things. 
https://go.dev/play/p/9hxWoDmvX2f

File: client/client.go (hunk 3 above)
Comment (lukanus) [reply client/client.go:253]: It's used twice below. so:
```
s := string(resp.Result)
[]byte(s[1 : len(s)-1])
```
vs 
```
[]byte(string(resp.Result)[1 : len(string(resp.Result))-1])
``` 
(assuming that's possible :D ) But yeah I probably migh used byte operations instead. the problem here is encoding so I used string

File: api/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (deelawn) [reply api/api.go:178]: Oh whoops. I didn't expand those lines in the diff. 👍

**Pull request merged**